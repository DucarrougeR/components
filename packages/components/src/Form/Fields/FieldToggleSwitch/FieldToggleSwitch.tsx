import React, { FunctionComponent, Ref } from 'react'
import styled, { StyledComponent } from 'styled-components'
import uuid from 'uuid/v4'
import { ComponentWithForm, withForm } from '../../Form'
import { ToggleSwitch, ToggleSwitchProps } from '../../Inputs/ToggleSwitch'
import { Field, FieldProps } from '../Field'

export interface FieldToggleSwitchProps extends FieldProps, ToggleSwitchProps {}
type ComponentType = FunctionComponent<FieldToggleSwitchProps>
type StyledComponentType = StyledComponent<
  ComponentType,
  FieldToggleSwitchProps
>

const InternalFieldToggleSwitch: ComponentType = (
  props: FieldToggleSwitchProps
) => {
  const {
    alignLabel,
    alignValidationMessage,
    label,
    labelWidth,
    labelFontWeight,
    validationMessage,
    id = uuid(),
    ...inputToggleSwitchProps
  } = props
  return (
    <Field
      id={id}
      alignLabel={alignLabel ? alignLabel : 'right'}
      {...props}
      alignValidationMessage={
        alignValidationMessage ? alignValidationMessage : 'bottom'
      }
    >
      <ToggleSwitch id={id} {...inputToggleSwitchProps} />
    </Field>
  )
}

const FieldToggleSwitchFactory = React.forwardRef<
  StyledComponentType,
  FieldToggleSwitchProps
>((props: FieldToggleSwitchProps, ref: Ref<StyledComponentType>) => (
  <InternalFieldToggleSwitch ref={ref} {...props} />
))

/** @component */
export const FieldToggleSwitch: StyledComponent<
  ComponentWithForm<FieldToggleSwitchProps>,
  FieldToggleSwitchProps
> = styled<ComponentWithForm<FieldToggleSwitchProps>>(
  withForm<FieldToggleSwitchProps>(FieldToggleSwitchFactory)
)``
