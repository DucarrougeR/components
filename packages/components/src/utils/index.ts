export * from './GlobalStyle'
export * from './useToggle'
