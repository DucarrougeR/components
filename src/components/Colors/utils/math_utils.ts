/**
 * Utility to convert a numerical value to a percent given the value and a max value.
 */
export const toPerc = (value: number, maxValue: number) =>
  Math.round((value / maxValue) * 100)
