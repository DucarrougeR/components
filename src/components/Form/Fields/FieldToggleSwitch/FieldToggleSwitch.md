The `<FieldToggleSwitch />` component is composed of a `<ToggleSwitch />`   component and a `<Label />` component. By default, the label will render to the right of the toggle switch, however this is adjustable with the alignLabel property.


```js
class ToggleSwitchParent extends React.Component {
  constructor(props) {
    super(props)
    this.state = { on: false }
  }

  handleChange(event) {
    this.setState({ on: event.target.checked })
  }

  render() {
    return (
      <FieldToggleSwitch
        label="Development Mode"
        alignLabel="left"
        onChange={this.handleChange.bind(this)}
        on={this.state.on}
      />
    )
  }
}

  <ToggleSwitchParent />
```
