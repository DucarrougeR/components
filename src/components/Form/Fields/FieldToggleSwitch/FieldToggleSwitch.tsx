import * as React from 'react'
import uuid from 'uuid/v4'
import { withForm } from '../../Form'
import { ToggleSwitch, ToggleSwitchProps } from '../../Inputs/ToggleSwitch'
import { Field, FieldProps } from '../Field'

interface FieldToggleSwitchProps extends FieldProps, ToggleSwitchProps {}

const InternalFieldToggleSwitch = (props: FieldToggleSwitchProps) => {
  const {
    alignLabel,
    alignValidationMessage,
    label,
    labelWidth,
    labelFontWeight,
    validationMessage,
    id = uuid(),
    ...inputToggleSwitchProps
  } = props
  return (
    <Field
      id={id}
      alignLabel={alignLabel ? alignLabel : 'right'}
      {...props}
      alignValidationMessage={
        alignValidationMessage ? alignValidationMessage : 'bottom'
      }
    >
      <ToggleSwitch id={id} {...inputToggleSwitchProps} />
    </Field>
  )
}

export const FieldToggleSwitch = withForm(InternalFieldToggleSwitch)
