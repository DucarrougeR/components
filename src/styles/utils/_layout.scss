// Returns a value from the $spacing list
@function spacing($key) {
  @return map-lookup($spacing, $key, '$spacing');
}
