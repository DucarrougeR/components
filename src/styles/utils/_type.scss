// Returns a value from the $test-size list
@function font-size($key) {
  @return map-lookup($text-size, $key, '$text-size');
 }

 // Returns a value from the $line-height list
 @function line-height($key) {
   @return map-lookup($line-height, $key, '$line-height');
 }

 @mixin fontSize($size) {
  font-size: $size;
  font-size: pxToRem($size);
}
