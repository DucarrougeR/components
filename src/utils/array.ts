/**
 * Returns an array of length `size` filled with zeros.
 * Useful for looping constructs where a length is known, but no array is provided.
 */

export const sizedArray = (size: number) => Array(size).fill(0)
