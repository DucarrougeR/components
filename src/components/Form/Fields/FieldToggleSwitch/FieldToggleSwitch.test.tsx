import 'jest-styled-components'
import * as React from 'react'
import { assertSnapshot } from '../../../../../test/utils/snapshot'
import { FieldToggleSwitch } from './FieldToggleSwitch'

test('A FieldToggleSwitch', () => {
  assertSnapshot(
    <FieldToggleSwitch
      label="👍"
      name="thumbsUp"
      id="thumbs-up"
      ariaId="thumb-toggle"
    />
  )
})

test('A FieldToggleSwitch with label aligned left', () => {
  assertSnapshot(
    <FieldToggleSwitch
      label="👍"
      name="thumbsUp"
      id="thumbs-up"
      ariaId="thumb-toggle"
      alignLabel="left"
    />
  )
})

test('A FieldToggleSwitch turned on', () => {
  assertSnapshot(
    <FieldToggleSwitch
      label="👍"
      name="thumbsUp"
      id="thumbs-up"
      on={true}
      ariaId="thumb-toggle"
    />
  )
})
