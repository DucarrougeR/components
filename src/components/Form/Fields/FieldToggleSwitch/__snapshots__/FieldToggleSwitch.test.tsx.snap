// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A FieldToggleSwitch 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row-reverse;
  -ms-flex-direction: row-reverse;
  flex-direction: row-reverse;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  margin-bottom: 0.5rem;
}

.c2 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  color: #262D33;
  font-size: 0.875rem;
  font-weight: 600;
  margin-left: 0.5rem;
  margin-right: 0.5rem;
  text-align: left;
  width: 20%;
}

.c3 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-bottom: 0.5rem;
}

.c4 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
}

.c5 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  cursor: pointer;
  display: inline-block;
  height: 1.25rem;
  position: relative;
  vertical-align: middle;
  width: 2.1875rem;
}

.c7 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  height: 100%;
  left: 0;
  opacity: 0;
  position: absolute;
  top: 0;
  width: 100%;
  z-index: 1;
}

.c9 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  border-radius: 1.25rem;
  bottom: 0;
  background-color: #C1C6CC;
  left: 0;
  position: absolute;
  right: 0;
  top: 0;
}

.c9:hover {
  box-shadow: 0 0 .01rem 0.01rem rgba(100,81,138,0.5);
}

.c11 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  border-radius: 50%;
  bottom: 0.125rem;
  background-color: #FFFFFF;
  height: 1rem;
  left: 0.125rem;
  position: absolute;
  width: 1rem;
}

.c10 {
  -webkit-transition: 300ms;
  transition: 300ms;
}

.c8 {
  -webkit-transition: 300ms;
  transition: 300ms;
}

.c6:focus + .c0 {
  box-shadow: 0 0 0 0.2rem rgba(100,81,138,0.4);
}

<div
  className="c0 c1"
>
  <label
    className="c0 c2"
    htmlFor="thumbs-up"
  >
    üëç
  </label>
  <div
    className="c0 c3"
  >
    <div
      className="c0 c4"
    >
      <div
        className="c0 c5"
        cursor="pointer"
      >
        <input
          className="c6 c0 c7"
          id="thumbs-up"
          name="thumbsUp"
          role="switch"
          type="checkbox"
        />
        <div
          className="c8 c0 c9"
        >
          <div
            className="c10 c0 c11"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`A FieldToggleSwitch turned on 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row-reverse;
  -ms-flex-direction: row-reverse;
  flex-direction: row-reverse;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  margin-bottom: 0.5rem;
}

.c2 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  color: #262D33;
  font-size: 0.875rem;
  font-weight: 600;
  margin-left: 0.5rem;
  margin-right: 0.5rem;
  text-align: left;
  width: 20%;
}

.c3 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-bottom: 0.5rem;
}

.c4 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
}

.c5 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  cursor: pointer;
  display: inline-block;
  height: 1.25rem;
  position: relative;
  vertical-align: middle;
  width: 2.1875rem;
}

.c7 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  height: 100%;
  left: 0;
  opacity: 0;
  position: absolute;
  top: 0;
  width: 100%;
  z-index: 1;
}

.c11 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  border-radius: 50%;
  bottom: 0.125rem;
  background-color: #FFFFFF;
  height: 1rem;
  left: 0.125rem;
  position: absolute;
  width: 1rem;
}

.c9 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  border-radius: 1.25rem;
  bottom: 0;
  background-color: #9785F2;
  left: 0;
  position: absolute;
  right: 0;
  top: 0;
}

.c9:hover {
  box-shadow: 0 0 .01rem 0.01rem rgba(100,81,138,0.5);
}

.c10 {
  -webkit-transform: translateX(0.9375rem);
  -ms-transform: translateX(0.9375rem);
  transform: translateX(0.9375rem);
  -webkit-transition: 300ms;
  transition: 300ms;
}

.c8 {
  -webkit-transition: 300ms;
  transition: 300ms;
}

.c6:focus + .c0 {
  box-shadow: 0 0 0 0.2rem rgba(100,81,138,0.4);
}

<div
  className="c0 c1"
>
  <label
    className="c0 c2"
    htmlFor="thumbs-up"
  >
    üëç
  </label>
  <div
    className="c0 c3"
  >
    <div
      className="c0 c4"
    >
      <div
        className="c0 c5"
        cursor="pointer"
      >
        <input
          aria-checked={true}
          checked={true}
          className="c6 c0 c7"
          id="thumbs-up"
          name="thumbsUp"
          role="switch"
          type="checkbox"
        />
        <div
          className="c8 c0 c9"
        >
          <div
            className="c10 c0 c11"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`A FieldToggleSwitch with label aligned left 1`] = `
.c3 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-bottom: 0.5rem;
}

.c4 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
}

.c5 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  cursor: pointer;
  display: inline-block;
  height: 1.25rem;
  position: relative;
  vertical-align: middle;
  width: 2.1875rem;
}

.c7 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  height: 100%;
  left: 0;
  opacity: 0;
  position: absolute;
  top: 0;
  width: 100%;
  z-index: 1;
}

.c9 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  border-radius: 1.25rem;
  bottom: 0;
  background-color: #C1C6CC;
  left: 0;
  position: absolute;
  right: 0;
  top: 0;
}

.c9:hover {
  box-shadow: 0 0 .01rem 0.01rem rgba(100,81,138,0.5);
}

.c11 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  border-radius: 50%;
  bottom: 0.125rem;
  background-color: #FFFFFF;
  height: 1rem;
  left: 0.125rem;
  position: absolute;
  width: 1rem;
}

.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  margin-bottom: 0.5rem;
}

.c2 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  color: #262D33;
  font-size: 0.875rem;
  font-weight: 600;
  margin-right: 0.5rem;
  text-align: right;
  width: 20%;
}

.c10 {
  -webkit-transition: 300ms;
  transition: 300ms;
}

.c8 {
  -webkit-transition: 300ms;
  transition: 300ms;
}

.c6:focus + .c0 {
  box-shadow: 0 0 0 0.2rem rgba(100,81,138,0.4);
}

<div
  className="c0 c1"
>
  <label
    className="c0 c2"
    htmlFor="thumbs-up"
  >
    üëç
  </label>
  <div
    className="c0 c3"
  >
    <div
      className="c0 c4"
    >
      <div
        className="c0 c5"
        cursor="pointer"
      >
        <input
          className="c6 c0 c7"
          id="thumbs-up"
          name="thumbsUp"
          role="switch"
          type="checkbox"
        />
        <div
          className="c8 c0 c9"
        >
          <div
            className="c10 c0 c11"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
